Fix invalid shard shapes from python api

jvegaTT opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/30634

### Problem description
core.py contains a function used to determine valid shard shapes which has two bugs
- It doesn't account for L1 alignment requirements on mesh shape
- It doesn't account for tile alignment requirements on the height dimension

### What's changed
Function was patched appropriately 

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/18565216613) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/18565254687) CI with demo tests passes (if applicable)
- [ ] [Sweeps Results](https://github.com/tenstorrent/tt-metal/actions/runs/18565342999)

jvegaTT committed:
Fix invalid shard shapes from python api

jvegaTT committed:
Applied AI suggestions

jvegaTT committed:
fixes spacing problems

bbradelTT commented on tests/sweep_framework/sweeps/ccl/generality/reduce_scatter.py:
```diff
@@ -247,10 +249,7 @@ def run(
-                eq, output = comp_equal(tt_output_tensor, ref)
-            else:
-                eq, output = comp_pcc(tt_output_tensor, ref)
+            eq, output = comp_pcc(tt_output_tensor, ref)
```
What is the reason for dropping the equality check?

jvegaTT commented on ttnn/ttnn/core.py:
```diff
@@ -254,6 +255,8 @@ def create_sharded_memory_config_(
         >>> tensor = ttnn.create_sharded_memory_config((5, 8), (320,64), ttnn.ShardStrategy.BLOCK, ttnn.ShardOrientation.ROW_MAJOR)
     """
+    L1_ALLIGNMENT = 16
+    TILE_ALLIGNMENT_REQUIREMENT = 32
```
WH and BH have the same L1 alignment requirements, however I agree its better to get this info from an API, however we are at the python level, seeing what the python HAL equivalent is

jvegaTT replied:
Seems like people are using magic number 16 all over the database. Core.py seems to be the most appropriate location for this info.

jvegaTT committed:
Added a common Alignment argument

jvegaTT committed:
Merge branch 'main' into jvega/fix_invalid_shard_shapes

amorrisonTT commented:
Are these changes covered by the [unit test](https://github.com/tenstorrent/tt-metal/blob/b207c557204fefc262e9ccb52ee53ecfed7fad59/tests/ttnn/unit_tests/operations/data_movement/test_core.py#L420) ? If not add some test cases.

jvegaTT committed:
Added tests and fixes to block sharded topology

jvegaTT commented:
> > I actually moved from test_create_sharded_memory_config to test_create_sharded_memory_config_ which was made to also handle tiled + I added bug fixes to it. There is no unit test for the second version. As per the docs, test_create_sharded_memory_config is to be replaced gradually to test_create_sharded_memory_config_
> 
> I think your changes need a unit test

I moved the test to the supported API from the deprecated one and added appropriate testing

jvegaTT committed:
Change DRAM_ALIGNMENT_WH to 32

Updated DRAM alignment width from 16 to 32.

jvegaTT committed:
Refactor alignment constants and update padding logic

jvegaTT committed:
Undoing dumb change AI made

TT-BrianLiu commented on ttnn/ttnn/core.py:
```diff
@@ -333,23 +341,53 @@ def create_sharded_memory_config_(
+                tensor_width_padded = (
+                    roundup(tensor_width, grid_size.x * TILE_ALLIGNMENT_REQUIREMENT)
+                    if tile_layout
+                    else roundup(tensor_width, grid_size.x * L1_ALIGNMENT)
```
Why is this needed? Shard shapes shouldn't need to think about this. The addresses of the shards themselves should be aligned.

Also, why is this only considering L1 alignment?

TT-BrianLiu commented on ttnn/ttnn/core.py:
```diff
@@ -254,6 +261,7 @@ def create_sharded_memory_config_(
     Example::
         >>> tensor = ttnn.create_sharded_memory_config((5, 8), (320,64), ttnn.ShardStrategy.BLOCK, ttnn.ShardOrientation.ROW_MAJOR)
     """
+    TILE_ALLIGNMENT_REQUIREMENT = 32
```
Is this number based on tile sizes or hardware alignment?

jvegaTT replied:
From the python layer when querying for a valid configuration, we don't have visibility of things like tiny tile, so we assume the largest possible

TT-BrianLiu reviewed (changes requested):
I have some concerns about the code, but also high-level:
- The ccl test infra seems really convoluted, wrapping layers and layers of function calls together to just get a valid memory config. I would suggest decoupling ccl infra from things like `ttnn. create_sharded_memory_config_`
- I don't think ttnn really maintains this API: `create_sharded_memory_config_` or `create_sharded_memory_config`. This bifurcation is really ugly and needs to be cleaned up.
